### ISSUE TITLE
pair_programming for git backend doesn't work right

### ISSUE THREAD

**Issue by imnitishng** [chaoss/grimoirelab-elk#825]
So I was exploring the Git enricher and came through this issue with  `pair_programming` set to `true` in `setup_cfg`.

The variables `m` and `n` set [here](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L414) through regex don't seem to work for the repository with multi-authored commits I've chosen (https://github.com/juliettepretot/npx-co-commit).

Based on the current logic both the variables are set to `None` because the regex used in `AUTHOR_P2P_REGEX` and `AUTHOR_P2P_NEW_REGEX` seems faulty, when they both should get populated by the 2 author names for that particular commit.

For instance let's look at this commit JSON data from Perceval 
```
{
    "backend_name": "Git",
    "backend_version": "0.12.0",
    "category": "commit",
    "classified_fields_filtered": null,
    "data": {
        "Author": "Juliette Pretot <hi@juliette.sh>",
        "AuthorDate": "Mon Oct 29 16:29:48 2018 +0100",
        "Commit": "Juliette Pretot <hi@juliette.sh>",
        "CommitDate": "Tue Nov 6 21:41:35 2018 +0100",
        "commit": "dbc4e2c18c893be4db75b472d4016e623316729e",
        "files": [
            {
                "action": "A",
                "added": "10",
                "file": ".gitignore",
                "indexes": [
                    "0000000",
                    "e45672e"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "24",
                "file": "README.MD",
                "indexes": [
                    "0000000",
                    "45d1af5"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "46",
                "file": "index.js",
                "indexes": [
                    "0000000",
                    "5f2990c"
                ],
                "modes": [
                    "000000",
                    "100755"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "27",
                "file": "package-lock.json",
                "indexes": [
                    "0000000",
                    "d0ef368"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "19",
                "file": "package.json",
                "indexes": [
                    "0000000",
                    "8f521a3"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "0"
            },
            {
                "action": "A",
                "added": "-",
                "file": "screenshot.gif",
                "indexes": [
                    "0000000",
                    "74de826"
                ],
                "modes": [
                    "000000",
                    "100644"
                ],
                "removed": "-"
            }
        ],
        "message": "initial commit\n\nCo-authored-by: mariiapunda <mariiapunda@users.noreply.github.com>",
        "parents": [],
        "refs": []
    },
    "origin": "https://github.com/juliettepretot/npx-co-commit",
    "perceval_version": "0.12.24",
    "search_fields": {
        "item_id": "dbc4e2c18c893be4db75b472d4016e623316729e"
    },
    "tag": "https://github.com/juliettepretot/npx-co-commit",
    "timestamp": 1585582796.576738,
    "updated_on": 1541536895.0,
    "uuid": "0c9876869058dc5b3f836365f3ce06a5b5575749"
}
```
According to the above commit data there must be 2 authors for the commit **Juliette Pretot** and **mariiapunda**  (based on the line `"message": "initial commit\n\nCo-authored-by: mariiapunda `)

But after fixing regex 
```
AUTHOR_P2P_REGEX = re.compile(r'(?P<first_authors>.* .*) and (?P<last_author>.* .*) (?P<email>.*)')
```
 to
```
AUTHOR_P2P_REGEX = re.compile(r'(?P<first_authors>.* ) (?P<last_author> .*) (?P<email>.*)')
```
what I observe is in the function `__get_authors()` [here](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L118) the committer name from this line 

```
"Author": "Juliette Pretot <hi@juliette.sh>",
```

is used and gets split into two names and saved in `item['data']['authors']` as `<class 'list'>: ['Juliette', 'Pretot']`.

AFAIK `item['data']['authors']` in this case should be `<class 'list'>: ['Juliette Pretot', 'mariiapunda']`.

I feel the logic being used here is wrong, from the regex to how multiple authors are being identified.
Please tell me if I am wrong with the repository I chose or the way I am interpreting the logic.

**Comment by imnitishng** [comment]
What I am thinking of is updating the regular expressions `AUTHOR_P2P_REGEX` and `AUTHOR_P2P_NEW_REGEX` to extract the names of both the committers correctly and then changing the logic used in `self.__get_authors()` method to return correct values to store in `item['data']['authors']`.
But the problem is there is no way to extract the real name of the second committer from 
```
 "message": "initial commit\n\nCo-authored-by: mariiapunda <mariiapunda@users.noreply.github.com>"
```

**Comment by imnitishng** [comment]
@valeriocos please have a look.

**Comment by valeriocos** [comment]
Hi @imnitishng I'm not sure I'll be able to check it today. Sorry for the inconvenience.

**Comment by imnitishng** [comment]
@valeriocos it's fine. Please have a look when you have the time to.

**Comment by imnitishng** [comment]
Hi @valeriocos sorry to bug, but can you test this so I can work on increasing coverage to 80%. 
Because tests don't seem to work for the  `pair_programming` parameter.

**Comment by valeriocos** [comment]
I'm on it @imnitishng , sorry for the late reaction

**Comment by valeriocos** [comment]
Thank you @imnitishng to have a look at this.

The problem should be due to the fact that both regexps (seem them below) assume that the author/committer have a first name and last name. In the example you report, there is a missing name (`mariiapunda <...>`)

```
AUTHOR_P2P_REGEX = re.compile(r'(?P<first_authors>.* .*) and (?P<last_author>.* .*) (?P<email>.*)')
AUTHOR_P2P_NEW_REGEX = re.compile(r"Co-authored-by:(?P<first_authors>.* .*)<(?P<email>.*)>\n?")
```

I have tried with the following conf/repo and the pair-programming works fine.
- setup.cfg
```
[git]
raw_index = git_cloudfoundry
enriched_index = git_cloudfoundry_enriched
pair-programming = true
latest-items = false
```
- projects.json
```
{
    "test": {
        "git": [
            "https://github.com/cloudfoundry/archiver"
        ]
    }
}
```

Perceval item:
```
{
    "backend_name": "Git",
    "backend_version": "0.12.0",
    "category": "commit",
    "classified_fields_filtered": null,
    "data": {
        "Author": "Atul Kshirsagar and James Myers <atul.kshirsagar@gmail.com>",
        "AuthorDate": "Tue Apr 28 13:56:05 2015 -0700",
        "Commit": "Atul Kshirsagar and James Myers <atul.kshirsagar@gmail.com>",
        "CommitDate": "Tue Apr 28 13:56:05 2015 -0700",
        "commit": "f598990ddfeda2767a09346acf844a7d216c1256",
        "files": [
            { ... }
        ],
        "message": "Change \u03a9 to Expect [#93246926]",
        "parents": [
            "8083f852b16af2fef833352bb8fe88dd2946f778"
        ],
        "refs": []
    },
    "origin": "https://github.com/cloudfoundry/archiver",
    "perceval_version": "0.13.0",
    "search_fields": {
        "item_id": "f598990ddfeda2767a09346acf844a7d216c1256"
    },
    "tag": "https://github.com/cloudfoundry/archiver",
    "timestamp": 1585912696.526519,
    "updated_on": 1430254565.0,
    "uuid": "0395e1c5acd0c33b61dbe91ac4f15c213a43ef59"
}
```

- regexps:


The regexp AUTHOR_P2P_REGEX is applied to the text below and the authors are correctly parsed
```
Atul Kshirsagar and James Myers <atul.kshirsagar@gmail.com>
```

**Comment by valeriocos** [comment]
To make sure that the new regexp works with should have test cases for both scenarios.

WRT increase the test coverage, I guess the easiest thing is to add/modify the test data for Git (https://github.com/chaoss/grimoirelab-elk/blob/master/tests/data/git.json) to have a commit with pair programming info.

WDYT?

**Comment by imnitishng** [comment]
Hi @valeriocos I get what you're saying, can you please clear up why `AUTHOR_P2P_NEW_REGEX`  exists then?
In the repository commit that I tested on ([here](https://github.com/juliettepretot/npx-co-commit/commit/c763cb96560bf30acf02f48205f731d92b910ca0)) the commit message states

> Enable users to pass flags
Co-authored-by: mariiapunda <mariiapunda@users.noreply.github.com>

And `AUTHOR_P2P_NEW_REGEX` extracts the author username from this line  which is present in `item['data']['message']`. (see first photo)

So I think the confusion is about which commits are classified as pair-programming commits as per the logic used here for [multi-author](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L417) and here for [multi-committer](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L424)

Can you please clear out which of the following commit according to ELK is classified with **multi-authors** and **multi-comitter**

![Screenshot from 2020-04-03 18-04-57](https://user-images.githubusercontent.com/35267629/78361347-e47b6f80-75d5-11ea-9a39-b6853873b97b.png)
[link to commit](https://github.com/juliettepretot/npx-co-commit/commits/master)

![Screenshot from 2020-04-03 18-05-00](https://user-images.githubusercontent.com/35267629/78361304-d2013600-75d5-11ea-8451-b365c392e4df.png)
[link to commit](https://github.com/cloudfoundry/archiver/commits/master?before=4ca7245c29b165a2c839a50fafde2c863cd955b1+35)

**Comment by valeriocos** [comment]
Hi @imnitishng ,

I undestand that the two regexps are used to catch author/committer information which may appear with the following shape:
- author1 and author2
- co-authored-by author

The cast to `set` and `list` is used to avoid duplicated authors: https://github.com/chaoss/grimoirelab-elk/blob/master/grimoire_elk/enriched/git.py#L132

`AUTHOR_P2P_NEW_REGEX` exists because user information can be appear as `Co-authored-by`

> Can you please clear out which of the following commit according to ELK is classified with multi-authors and multi-comitter

I understand that both cases aren't considered as pair-programming for ELK, since the `Author` and `Commit` fields don't match the regexp and there is no `co-authored` info. WDYT?

```
{
    "backend_name": "Git",
    "backend_version": "0.12.0",
    "category": "commit",
    "classified_fields_filtered": null,
    "data": {
        "Author": "Juliette Pretot <hi@juliette.sh>",
        "AuthorDate": "Fri Nov 9 16:45:23 2018 +0100",
        "Commit": "Juliette Pretot <hi@juliette.sh>",
        "CommitDate": "Fri Nov 9 16:47:33 2018 +0100",
        "commit": "c763cb96560bf30acf02f48205f731d92b910ca0",
        "files": [ ... ],
        "message": "Enable users to pass flags\n\nCo-authored-by: mariiapunda <mariiapunda@users.noreply.github.com>",
        "parents": [
            "58e2655322a2b0df1263695b8efb39f7158e5821"
        ],
        "refs": []
    },
    "origin": "https://github.com/juliettepretot/npx-co-commit",
    "perceval_version": "0.13.0",
    "search_fields": {
        "item_id": "c763cb96560bf30acf02f48205f731d92b910ca0"
    },
    "tag": "https://github.com/juliettepretot/npx-co-commit",
    "timestamp": 1585929173.486108,
    "updated_on": 1541778453.0,
    "uuid": "e0d165308041e08cf02992d6e9c1f79cffaea6b9"
}
```

```
{
    "backend_name": "Git",
    "backend_version": "0.12.0",
    "category": "commit",
    "classified_fields_filtered": null,
    "data": {
        "Author": "John Shahid <jvshahid@gmail.com>",
        "AuthorDate": "Mon Feb 6 14:16:40 2017 -0500",
        "Commit": "Danny Cao <dcao@us.ibm.com>",
        "CommitDate": "Mon Feb 6 14:16:40 2017 -0500",
        "Signed-off-by": [
            "Danny Cao <dcao@us.ibm.com>"
        ],
        "commit": "59f15fd17404dceccb25d8b62df1271035b707a5",
        "files": [ ... ],
        "message": "Enforce canonical import path\n\n[#123291791]\n\nSigned-off-by: Danny Cao <dcao@us.ibm.com>",
        "parents": [
            "b15977cce944c4a672a7c0f45f5264608e615e61"
        ],
        "refs": []
    },
```

**Comment by imnitishng** [comment]
Yes.

>  there is no co-authored info.

I think there might be a change since this implementation because now `co-authored` info is available in `item[data][message]`, and not in `item[data][Author]` based on [this](https://github.com/chaoss/grimoirelab-elk/blob/8ea3455bd639de695e1a4d2f56ed86df4e5818c7/grimoire_elk/enriched/git.py#L415) logic.

Anyway, thank you very much, I will send a PR with added data for pair-programming in [git.json](https://github.com/chaoss/grimoirelab-elk/blob/master/tests/data/git.json) soon.

**Comment by valeriocos** [comment]
@imnitishng can we close this issue or there is something left to do (maybe improving the documentation)? 

I understand that the PR https://github.com/chaoss/grimoirelab-elk/pull/832 shows that the pair programming study works fine

**Comment by imnitishng** [comment]
Hi @valeriocos 
I'm sorry I don't see the places where documentation might need improvement. 
If there is work to do then please guide me I'll finish it, other than that I think the issue can be closed.

**Comment by valeriocos** [comment]
Hi @imnitishng , thank you for the quick answer!

I would improve the docstring at https://github.com/chaoss/grimoirelab-elk/blob/master/grimoire_elk/enriched/git.py#L390 to explain the overall process to identify and handle pair programmed commits. The docstring could also include snippets of Perceval documents to let the developer understand which commits are considered as pair programmed and which are not. WDYT?

**Comment by imnitishng** [comment]
Okay sure, I will do it and update the PR. Thanks.

**Comment by imnitishng** [comment]
Hi @valeriocos, have a look at the PR, tell about any changes you want to add, else we can close the issue.

**Comment by valeriocos** [comment]
Thank you @imnitishng , I left some comments. Please consider to edit the body of the pull request with `Fixes #825`. So, once the PR is merged, this issue will be automatically close.
